File: src/lib/api/favorites.ts
import { invoke } from '@tauri-apps/api/core';
import type { FavoriteChannel, WatchHistoryItem } from '$lib/types';

/**
 * Toggle a channel's favorite status.
 * @param channelId - The UUID of the channel
 * @param playlistId - The UUID of the playlist
 * @param favoriteType - The type of favorite ('channel' or 'global')
 * @returns True if the channel is now favorited, false if unfavorited
 */
export async function toggleFavorite(
  channelId: string,
  playlistId: string,
  favoriteType: string
): Promise<boolean> {
  return await invoke<boolean>('toggle_favorite', {
    channelId,
    playlistId,
    favoriteType,
  });
}

/**
 * Get all favorite channels.
 * @param playlistId - Optional playlist ID to filter favorites by
 * @returns Array of FavoriteChannel objects
 */
export async function getFavorites(playlistId?: string): Promise<FavoriteChannel[]> {
  return await invoke<FavoriteChannel[]>('get_favorites', { playlistId });
}

/**
 * Persist a user drag-reorder of a playlist's favorites.
 * @param playlistId - The UUID of the playlist
 * @param orderedChannelIds - Favorited channel IDs in their new front-to-back order
 */
export async function reorderFavorites(playlistId: string, orderedChannelIds: string[]): Promise<void> {
  await invoke<void>('reorder_favorites', {
    playlistId,
    orderedChannelIds,
  });
}

/**
 * Get recently watched channels.
 * @param limit - Maximum number of items to return
 * @returns Array of WatchHistoryItem objects
 */
export async function getRecentlyWatched(limit: number): Promise<WatchHistoryItem[]> {
  return await invoke<WatchHistoryItem[]>('get_recently_watched', { limit });
}

/**
 * Save the playback position for a channel.
 * @param channelId - The UUID of the channel
 * @param playlistId - The UUID of the playlist
 * @param positionSeconds - The current playback position in seconds
 * @param totalSeconds - The total duration of the content in seconds
 */
export async function savePlaybackPosition(
  channelId: string,
  playlistId: string,
  positionSeconds: number,
  totalSeconds: number
): Promise<void> {
  await invoke<void>('save_playback_position', {
    channelId,
    playlistId,
    positionSeconds,
    totalSeconds,
  });
}

/**
 * Remove a single entry from watch history.
 * @param id - The UUID of the watch history row
 */
export async function removeWatchHistoryItem(id: string): Promise<void> {
  await invoke<void>('remove_watch_history_item', { id });
}

/**
 * Permanently delete all watch history entries.
 */
export async function clearWatchHistory(): Promise<void> {
  await invoke<void>('clear_watch_history');
}

File: src-tauri/src/commands/favorites.rs
use crate::db;
use crate::error::{CommandError, CommandResult};
use crate::state::AppState;
use crate::types::{FavoriteChannel, FavoriteType, WatchHistoryItem};
use tauri::State;

#[tauri::command]
pub async fn toggle_favorite(
    state: State<'_, AppState>,
    channel_id: String,
    playlist_id: String,
    favorite_type: String,
) -> CommandResult<bool> {
    let favorite_type = FavoriteType::from_str(&favorite_type)
        .ok_or_else(|| CommandError::Api(format!("Invalid favorite type: {favorite_type}")))?;
    db::with_conn(&state.db, move |conn| {
        Ok(db::favorites::toggle(conn, &channel_id, &playlist_id, favorite_type)?)
    })
    .await
}

#[tauri::command]
pub async fn get_favorites(
    state: State<'_, AppState>,
    playlist_id: Option<String>,
) -> CommandResult<Vec<FavoriteChannel>> {
    db::with_conn(&state.db, move |conn| {
        Ok(db::favorites::list(conn, playlist_id.as_deref())?)
    })
    .await
}

#[tauri::command]
pub async fn reorder_favorites(
    state: State<'_, AppState>,
    playlist_id: String,
    ordered_channel_ids: Vec<String>,
) -> CommandResult<()> {
    db::with_conn(&state.db, move |conn| {
        Ok(db::favorites::reorder(conn, &playlist_id, &ordered_channel_ids)?)
    })
    .await
}

#[tauri::command]
pub async fn get_recently_watched(
    state: State<'_, AppState>,
    limit: i64,
) -> CommandResult<Vec<WatchHistoryItem>> {
    db::with_conn(&state.db, move |conn| Ok(db::favorites::recently_watched(conn, limit)?)).await
}

#[tauri::command]
pub async fn save_playback_position(
    state: State<'_, AppState>,
    channel_id: String,
    playlist_id: String,
    position_seconds: i64,
    total_seconds: i64,
) -> CommandResult<()> {
    db::with_conn(&state.db, move |conn| {
        if !db::read_settings(conn).track_watch_history {
            return Ok(());
        }
        Ok(db::favorites::save_playback_position(
            conn,
            &channel_id,
            &playlist_id,
            position_seconds,
            total_seconds,
        )?)
    })
    .await
}

#[tauri::command]
pub async fn remove_watch_history_item(state: State<'_, AppState>, id: String) -> CommandResult<()> {
    db::with_conn(&state.db, move |conn| Ok(db::favorites::remove_history_item(conn, &id)?)).await
}

#[tauri::command]
pub async fn clear_watch_history(state: State<'_, AppState>) -> CommandResult<()> {
    db::with_conn(&state.db, |conn| Ok(db::favorites::clear_history(conn)?)).await
}
